### PR TITLE
Support Personal access token method for Tableau Connection

### LINKED ISSUE #25904: Support Personal access token method for Tableau Connection
### Description

In Tableau Online, you can use user name and password method for REST API calls, [only if you don't use MFA](https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api_concepts_auth.htm).
Thus, organizations that enabled MFA in Tableau, have to use [Personal Access Tokens](https://help.tableau.com/current/online/en-us/security_personal_access_tokens.htm) for API calls to Tableau.
However, [Airflow deprecated the usage of Personal Access Tokens](https://github.com/apache/airflow/pull/16916) in [airflow.providers.tableau.hooks.tableau](https://airflow.apache.org/docs/apache-airflow-providers-tableau/stable/_modules/airflow/providers/tableau/hooks/tableau.html), since [there is a problem](https://github.com/apache/airflow/issues/16669) that each time you try to open a parallel connection, the previous one would be invalidated.

### Use case/motivation

I would like to use the Personal Access Token method for API calls to Tableau.
I see 2 solutions that might be applicable:
1. Create 1 connection to Tableau and use that for all API calls from Airflow in a specific timeframe. I am not sure if this is feasible
2. Create a pool of tokens and for each connection to Tableau from Airflow, create a new connection id with 1 of the unused tokens. I saw that on Airflow, you can [generate connection on the fly](https://airflow.apache.org/docs/apache-airflow/1.10.14/howto/connection/index.html#generating-a-connection-uri). I am not sure how to handle the token pool though

I am open to other solutions and guidance. Also, I am not sure I can submit a PR since I have never contributed to an open source project but I can try.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by baruch-shteken**
closes: #25904

I read the [Tableau Python SDK docs](https://tableau.github.io/server-client-python/docs/api-ref#personalaccesstokenauth-class) and I didn't see any meaningful feature that can help solving this problem.
I decided that what I can do changing the connection setting in the Tableau hook.
Instead of generating the connection 1 time when the hook is initialized ([enter function](https://github.com/apache/airflow/blob/main/airflow/providers/tableau/hooks/tableau.py#L98)), I would like the tableau connection to be set before any API request to Tableau. This should solve the problem because the Tableau session_token will be newly generated before every call.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by Michae94** [comment]
Hey @baruch-shteken , thanks for posting a PR for this issue.     
I was wondering how this new setup will work with multiple concurrent tasks using the same PAT.      

From my understanding, if a job try to authenticate using the same PAT as a concurrent job, the former job will fail. 

Am I missing something ?

**Comment by baruch-shteken** [comment]
Thanks @Michae94 for your question.
**This is my understanding of how PATs work**:
When a PAT is used for an authentication, it creates a new `session_token` which is then used in subsequent REST API calls. When the same PAT is used to authenticate again, it creates a new `session_token` and invalidates the previous `session_token`. ([Tableau docs](https://help.tableau.com/current/online/en-us/security_personal_access_tokens.htm))
**This is my understanding of how Airflow works**:
For the first call to Tableau in a specific task, we use the PAT to authenticate to tableau. This generates a `session_token`. We can't reuse the `session_token` that was already created in a different task. (I also checked [the library](https://tableau.github.io/server-client-python/docs/api-ref#personalaccesstokenauth-class) Airflow uses to send Tableau requests and didn't find anything useful we can use). 
**Example of the problem**:
Let's take an example of a DAG that has 2 parallel tasks that refresh 2 Tableau data sources. One of these tasks will generate and use a `session_token` to refresh the data source. The other task will do the same and also invalidate the first task `session_token`. When the first task polls Tableau to check if the refresh finished, it will fail because the `session_token` changed.

From the above, I drew the conclusion that if we authenticate every-time to Tableau for every call, the process should work. That being said, I didn't test it end2end. Should I test it with my private Tableau data sources? Can I create a test for it here?

**Comment by Michae94** [comment]
Ok now I understand better your proposition. 
Potentially, it means that this improvement will not be usable for Airflow kubernetes workers, as each task is performed on a different pod (in this context, ~machine).
Thanks for explaning it 🙏

**Comment by baruch-shteken** [comment]
I don't know how Airflow kubernetes workers work but if they all use the code I propose here and even with the same PAT, it should still be ok. Because each task will authenticate again before sending a request to Tableau.

**Comment by eladkal** [comment]
> Example of the problem:
Let's take an example of a DAG that has 2 parallel tasks that refresh 2 Tableau data sources. One of these tasks will generate and use a session_token to refresh the data source. The other task will do the same and also invalidate the first task session_token. When the first task polls Tableau to check if the refresh finished, it will fail because the session_token changed

If this is true then it will never really work regardless of the underlying archetacture (Celery, Kubernetes, etc..)

The question that needs to be answered is how Tableau recommands to do authentication from two seperated proccesses that tries to refresh two different dashboards at the same time and they are not running as part of the same proccess.

**Comment by baruch-shteken** [comment]
I tested my changes locally (I didn't manage to pull my changes from GitHub. However, we use an Airflow docker so I changed the file `airflow/providers/tableau/hooks/tableau.py` in the docker to my new code).
**Tl;dr** the code change did solve some of the problem but not all of it.
**Explanation**: The set up is 6 data source refresh tasks. Each task triggers the data source refresh and every 3 minutes sends a request to Tableau to check if the refresh finished. When I triggered the DAG, all tasks excepts 1 failed. I remembered that there is a problem of concurrency because if all tasks try to get a new session token at the same time, only 1 will succeed. So I implemented a 2 seconds delay between the tasks in the DAG file. After this change, all the tasks finished successfully.
This means that as long as only 1 task requests a session token at any moment, the process will work.

@Michae94 Do you have any recommendation for me on how to improve the process? Can I configure somewhere in the Airflow code that only 1 Tableau hook object can be active at any moment?
@eladkal I can ask Tableau their recommendation about authenticating from two separated processes that try to refresh two different dashboards at the same time but they need to use the same PAT.

**Comment by eladkal** [comment]
> @Michae94 Is the concurrency problem what you meant before in Kubernetes? Do you have any recommendation for me on how to improve the process? Can I configure somewhere in the Airflow code that only 1 Tableau hook object can be active at any moment?

You can not. This can only be enforced from users side if they create pool for tableau and don't allow more than 1 tableau task to run at the same time.

>@eladkal I can ask Tableau their recommendation about authenticating from two separated processes that try to refresh two different dashboards at the same time but they need to use the same PAT.

Sure.

From Airflow point of view - we only can enable capabilities of Tableau SDK so if the SDK is missing a feature that should be a feature request to Tableau.
